iree/compiler/Dialect/HAL/Target/TargetBackend.h

```diff
@@ -0,0 +1,305 @@
+  // The pass manager will be nested on `hal.executable` such that the pipeline
+  // will only run on executable contents.
+  //
+  // Backend transformation passes must check that the source op they receive
+  // is for them using the `target_backend` attribute. Backends may have
+  // multiple source ops in the same executable to transform such as when
+  // multiple target configurations are requested. Use the `matchPattern`
+  // utility when comparing the target backend name.
+  //
+  // For example, as input:
+  //   hal.executable @some_executable {
+  //     hal.interface @main_io {
+  //       hal.interface.binding @arg0, set=0, binding=0, ...
+  //       hal.interface.binding @arg1, set=0, binding=1, ...
+  //     }
+  //     hal.executable.entry_point @main attributes {
+  //       interface = @main_io,
+  //       ordinal = 0 : i32,
+  //       signature = (tensor<4xf32>) -> tensor<4xf32>
+  //     }
+  //     hal.executable.target "target-backend" {
+  //       module { ... }
+  //     }
+  //   }
+  //
+  // As output:
+  //   hal.executable @some_executable {
+  //     hal.interface @main_io ...
+  //     hal.executable.entry_point @main ...
+  //     hal.executable.target "target-backend" {
+  //       module { spv.module { ... } }
+  //     }
+  //   }
+  // TODO(benvanik): migrate this to the dynamic pipeline pass infra when it
+  // exists. This will likely change to be a function that registers handlers
+  // for target-specific name, attributes, etc. For now the executable target
+  // is passed to allow snooping.
+  virtual void buildTranslationPassPipeline(
+      IREE::HAL::ExecutableTargetOp targetOp, OpPassManager &passManager) = 0;
+
+  // TODO(benvanik): define linkage rules.
+  // Major thing to figure out here is how to rewrite the executable references.
+  // We may want to move executable selection into the hal.device.switch of the
+  // dispatches so that they can be more easily replaced per-target.
+  virtual LogicalResult linkExecutables(mlir::ModuleOp moduleOp) {
+    return success();
+  }
+
+  // Serializes the given |targetOp| executable produced by this backend to one
+  // or more binary byte buffer formats used for storage in the module file.
+  // Implementations should insert `hal.executable.binary` ops for each format
+  // (such as x64 and arm64 for compiled LLVM blobs, etc).
+  //
+  // If no serialization is provided then lowering the parent module into a
+  // binary format (such as to the IREE VM) will fail.
+  virtual LogicalResult serializeExecutable(
+      IREE::HAL::ExecutableTargetOp targetOp, OpBuilder &executableBuilder) {
+    llvm_unreachable("unimplemented serializeExecutable");
+    return failure();
+  }
+
+ protected:
+  // Calculates the workgroup size (x, y, z).
+  virtual std::array<Value, 3> calculateDispatchWorkgroupSize(
+      Location loc, IREE::HAL::ExecutableOp executableOp,
+      IREE::HAL::ExecutableEntryPointOp entryPointOp, Value workload,
```

Comment:
The problem is that if we fix a single shaped workload that implies that we have a single shaped output and we want dispatch regions to be able to have many outputs. That's how things were for awhile but I wanted to remove that limitation: the workload now just denotes how much work is to be done but not the shape of the work. The shape should be based on the actual shape of the I/O and operations, and all of those are available here (entryPointOp.signature has the original tensor<> signature). That gives the backends more control over how they interpret shapes and compute workgroup counts (whether they need to be in tiles or in items, how bounds checking/padding happens, etc).

For dynamic shapes we also don't know the exact shapes (or maybe anything but the rank) when first translating the executable and even if we could wait until we saw the dispatch that's not enough: thanks to dynamic shapes every single dispatch could differ so there's no data we should be relying on for correctness there (just potential specialization info).

The nice thing is that only the default recordDispatch implementation calls these functions; if you are doing your own recordDispatch then you can get access to the exact shapes of all I/O (even if dynamic), though by the point you are recording dispatches you've likely already fixed your workgroup size.

I left some TODOs in here with notes of where you may want to change this in SPIR-V-land: https://github.com/google/iree/blob/d1db8b4721cc50c70eb5be2007e7c2a1a1d181f7/iree/compiler/Dialect/HAL/Target/VulkanSPIRV/VulkanSPIRVTarget.cpp#L167-L179

This will be useful when we want to support runtime-chosen workgroup sizes as specialization constants - but for now this is at least an improvement over one workgroup size for all backends and all dispatches :)

TL;DR: a backend can do whatever it wants now but there are no assumptions made in IREE about the shape of work or how that work is tiled. Linalg-on-tensors as a tiling mechanism running prior to IREE may be able to do much smarter things and then just pass around the workload for IREE to schedule things, but the final linalg-to-blah lowering could still access the info from the original linalg-on-tensor pass that includes shapes.

Reply:
Also, what this frees you up for is looking at the actual ops inside the dispatch region to compute the value: is it a depthwise conv and what are the best values for that, etc - even if there's a reshape at the end fused in that changes the shape and would otherwise mess up logic. Some kind of cost modeling that is based on the context of the work being performed will let you do a lot of good things (or, learn them with magic ML 🦄  ;)